### PR TITLE
Block toss

### PR DESCRIPTION
**Pull request by ezrn**
## Describe your changes
- Improved the look of block toss and its growing mechanic
- Made it a bit smaller(harder to hit) but more frequent(lower cooldown)
- Aoe damage now works separately from direct hit damage(values can be adjusted separately)
## Checklist before requesting a review
- [X] I have performed a self-review of my code
- [X] I have tested my changes.

### CODE REVIEW

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java [hunk 1]
```diff
@@ -92,7 +97,7 @@ public void spawn(double size) {
         // Create block displays
         displayBlocks = new ArrayList<>();
         final World world = location.getWorld();
-        for (int i = 0; i < 10; i++) { // 10 block displays
+        for (int i = 0; i < 30; i++) { // 10 block displays
```
Comment (braulio-dev) [inline champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java:100]: This can cause some unnecessary overhead, are you sure creating 30 display blocks is necessary?

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java [hunk 2]
```diff
@@ -101,14 +106,21 @@ public void spawn(double size) {
 
             // Get a random translation so the boulder looks more natural
             // Center the block display by subtracting 0.5, acts as our reference point for offsets
-            float xTranslation = (float) ((Math.random()) - 0.5f - 0.5f);
-            float yTranslation = (float) ((Math.random()) - 0.5f) - 0.5f;
-            float zTranslation = (float) ((Math.random()) - 0.5f - 0.5f);
+            float xTranslation = (float) ((Math.random()) - 0.5f) * displacement;
+            float yTranslation = (float) ((Math.random()) - 0.5f) * displacement;
+            float zTranslation = (float) ((Math.random()) - 0.5f) * displacement;
```
Comment (braulio-dev) [inline champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java:111]: This actually isn't a displacement, this makes sure that the displacement is absolute and not relative. Display blocks' locations are shifted by 0.5 in each direction, so your effective displacement here is `displacement + 0.5` instead of just `displacement`.

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java [hunk 3]
```diff
@@ -329,6 +337,19 @@ public void impact(Player caster) {
         impactLocation.getWorld().playSound(impactLocation, Sound.ENTITY_ZOMBIE_BREAK_WOODEN_DOOR, 0.3f, 1.5f);
     }
 
+    private void doDamage(Location impactLocation, LivingEntity ent, Player caster, Boolean isDirect){
```
Comment (braulio-dev) [inline champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java:340]: `Boolean` instead of `boolean`

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java [hunk 4]
```diff
@@ -309,15 +320,12 @@ public void impact(Player caster) {
...
-                Vector knockback = ent.getLocation().toVector().subtract(impactLocation.toVector());
-                final double strength = (radius * radius - ent.getLocation().distanceSquared(impactLocation)) / (radius * radius);
-                VelocityData velocityData = new VelocityData(knockback, strength, false, 0.0, 0.0, 3.0, true);
-                UtilVelocity.velocity(ent, caster, velocityData);
-                UtilDamage.doCustomDamage(new CustomDamageEvent(ent, caster, null, EntityDamageEvent.DamageCause.CUSTOM, damage, false, skill.getName()));
-                UtilMessage.simpleMessage(ent, skill.getName(), "<alt2>%s</alt2> hit you with <alt>%s</alt>.", caster.getName(), skill.getName());
+                if(!ent.hasLineOfSight(impactLocation)) continue;
+                if(!alreadyHit) {
+                    damaged.add(ent);
+                    doDamage(impactLocation, ent, caster, false);
```
Comment (braulio-dev) [inline champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java:327]: No point on looping through the entire array if you're only going to hit one entity. You also shouldn't need this loop or the entire logic at all if you did direct impact logic properly to get nearby entities. Unnecessary overhead.

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java [hunk 5]
```diff
@@ -269,14 +273,21 @@ private void playCues() {
...
         }
 
+        boolean alreadyHit = false;
+
         this.impacted = true;
         final Location impactLocation = getCenterLocation();
 
+        if (target != null) {
+            doDamage(impactLocation, target, caster, true);
+            alreadyHit = true;
```
Comment (braulio-dev) [inline champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java:288]: This is going to cause the attacker players to not receive feedback chat messages, can you verify this?

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java [hunk 6]
```diff
@@ -329,6 +337,19 @@ public void impact(Player caster) {
...
+        Vector knockback = ent.getLocation().toVector().subtract(impactLocation.toVector());
+        final double strength = (radius * radius - ent.getLocation().distanceSquared(impactLocation)) / (radius * radius);
+        VelocityData velocityData = new VelocityData(knockback, strength, false, 0.0, 0.0, 3.0, true);
+        UtilVelocity.velocity(ent, caster, velocityData);
+        if(isDirect) {
+            UtilDamage.doCustomDamage(new CustomDamageEvent(ent, caster, null, EntityDamageEvent.DamageCause.CUSTOM, damage, false, skill.getName()));
+        } else {
+            UtilDamage.doCustomDamage(new CustomDamageEvent(ent, caster, null, EntityDamageEvent.DamageCause.CUSTOM, damage / 2, false, skill.getName()));
+        }
+        UtilMessage.simpleMessage(ent, skill.getName(), "<alt2>%s</alt2> hit you with <alt>%s</alt>.", caster.getName(), skill.getName());
```
Comment (braulio-dev) [inline champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java:350]: If you have this here remove lines 332-336

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java [hunk 7]
```diff
@@ -309,15 +320,12 @@ public void impact(Player caster) {
...
                 // Damage anybody who is not friendly
-                damaged.add(ent);
-                Vector knockback = ent.getLocation().toVector().subtract(impactLocation.toVector());
-                final double strength = (radius * radius - ent.getLocation().distanceSquared(impactLocation)) / (radius * radius);
-                VelocityData velocityData = new VelocityData(knockback, strength, false, 0.0, 0.0, 3.0, true);
-                UtilVelocity.velocity(ent, caster, velocityData);
-                UtilDamage.doCustomDamage(new CustomDamageEvent(ent, caster, null, EntityDamageEvent.DamageCause.CUSTOM, damage, false, skill.getName()));
-                UtilMessage.simpleMessage(ent, skill.getName(), "<alt2>%s</alt2> hit you with <alt>%s</alt>.", caster.getName(), skill.getName());
+                if(!ent.hasLineOfSight(impactLocation)) continue;
+                if(!alreadyHit) {
```
Comment (braulio-dev) [inline champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java:325]: Just remove the directly impacted entity from `nearby` so you don't have to check if the entities are already hit.

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java [hunk 8]
```diff
@@ -297,7 +308,7 @@ public void impact(Player caster) {
             UtilVelocity.velocity(vehicle, caster, velocityData);
         }
 
-        // Damage and heal
+        // Damage
         final List<KeyValue<LivingEntity, EntityProperty>> nearby = UtilEntity.getNearbyEntities(caster, impactLocation, radius, EntityProperty.ALL);
         if (caster.getLocation().distanceSquared(impactLocation) <= radius * radius) {
             nearby.add(new KeyValue<>(caster, EntityProperty.FRIENDLY));
```
Comment (braulio-dev) [inline champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java:314]: Remove this if you removed the healing effect

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java [hunk 9]
```diff
@@ -297,7 +308,7 @@ public void impact(Player caster) {
             UtilVelocity.velocity(vehicle, caster, velocityData);
         }
 
-        // Damage and heal
+        // Damage
         final List<KeyValue<LivingEntity, EntityProperty>> nearby = UtilEntity.getNearbyEntities(caster, impactLocation, radius, EntityProperty.ALL);
```
Comment (braulio-dev) [inline champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java:312]: Replace `EntityProperty.ALL` with `EntityProperty.ENEMY` if you removed the healing factor.

File: champions/src/main/resources/configs/skills/skills.yml [hunk 10]
```diff
@@ -590,19 +590,8 @@ skills:
...
-      baseDamage: 4.0
-      damageIncreasePerLevel: 1.0
-      baseRadius: 4.0
-      radiusIncreasePerLevel: 0.5
-      baseSpeed: 1.4
-      speedIncreasePerLevel: 0.1
-      hitBoxSize: 1.0
-      size: 0.6
-      sizePerLevel: 0.2
+      cooldown: 6.0
```
Comment (braulio-dev) [inline champions/src/main/resources/configs/skills/skills.yml:593]: All these should be here.

```
        baseRadius = getConfig("baseRadius", 3.0, Double.class);
        radiusIncreasePerLevel = getConfig("radiusIncreasePerLevel", 0.0, Double.class);
        baseSpeed = getConfig("baseSpeed", 1.5, Double.class);
        speedIncreasePerLevel = getConfig("speedIncreasePerLevel", 0.0, Double.class);
        size = getConfig("size", 0.5, Double.class);
        sizePerLevel = getConfig("sizePerLevel", 0.0, Double.class);
        hitBoxSize = getConfig("hitBoxSize", 1.5, Double.class);
        startSize = getConfig("startSize", 0.1, Double.class);
        displacement = getConfig("displacement", 0.25, Double.class);

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/sword/BlockToss.java [hunk 11]
```diff
@@ -285,11 +305,13 @@ public void collisionCheck() {
                         continue;
                     }
                 } else if (boulder.isThrown()) {
-                    final List<Entity> nearby = referenceEntity.getNearbyEntities(hitBoxSize, hitBoxSize, hitBoxSize);
+                    final List<LivingEntity> nearby = UtilEntity.getNearbyEnemies(caster, referenceEntity.getLocation(), hitBoxSize);
                     nearby.remove(caster);
-                    nearby.removeIf(entity -> !(entity instanceof LivingEntity) || entity instanceof ArmorStand);
+
```
Comment (braulio-dev) [inline champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/sword/BlockToss.java:310]: Why did you remove the logic that prevents other entities and armor stands from being damaged? This will cause BlockToss to damage Item Frames, etc

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/sword/BlockToss.java [hunk 12]
```diff
@@ -244,6 +261,9 @@ public void onHit(ProjectileHitEvent event) {
         if (!(event.getEntity() instanceof Arrow arrow) || !(arrow.getShooter() instanceof Player player)) {
             return;
         }
+        if(event.getHitEntity() != null){
+            if(!(event.getHitEntity() instanceof LivingEntity)) return;
+        }
```
Comment (braulio-dev) [inline champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/sword/BlockToss.java:266]:
```suggestion
        if (!(event.getHitEntity() instanceof LivingEntity livingEntity)) {
            return;
        }
```

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/sword/BlockToss.java [hunk 13]
```diff
@@ -220,7 +231,13 @@ public void updateCharge() {
...
 
                 if (chargeData.getCharge() < 1) {
-                    chargeData.boulder.setSize(chargeData.boulder.getSize() + getSize(level) / 20);
+                    float chargeProgress = chargeData.getCharge(); // 0 to 1
+
+                    double newSize = startSize + chargeProgress * (size - startSize);
+                    chargeData.boulder.setSize(newSize);
+
+                    float newDisplacement = 0.05F + chargeProgress * (0.4F - 0.05F);
+                    chargeData.boulder.setDisplacement(newDisplacement);
```
Comment (braulio-dev) [inline champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/sword/BlockToss.java:240]: Correct me if I'm wrong, but this isn't actually working because the display entities' displacement is not being updated, but instead it's only queried when they are spawned.

**Review by braulio-dev — CHANGES_REQUESTED** [review]

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java (hunk 2 above)
Comment (ezrn) [reply champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java:111]: this is OK, that was my intention and it works as I desired

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/sword/BlockToss.java (hunk 13 above)
Comment (ezrn) [reply champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/sword/BlockToss.java:240]: no it updates

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java (hunk 5 above)
Comment (ezrn) [reply champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java:288]: yes because I forgot to add them to the damaged map

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java (hunk 7 above)
Comment (ezrn) [reply champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java:325]: since I already use this boolean system in previous parts of the logic without access to the nearby map that contains direct hit player, id like to keep it here.

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java (hunk 4 above)
Comment (ezrn) [reply champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java:327]: The rest of the loop logic doesnt activate unless there are players nearby?

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java (hunk 6 above)
Comment (ezrn) [reply champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java:350]: the previous damage line only prints for the user, not the person getting hit

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/sword/BlockToss.java (hunk 11 above)
Comment (ezrn) [reply champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/sword/BlockToss.java:310]: because it now only checks for livingentity not entity

File: champions/src/main/resources/configs/skills/skills.yml (hunk 10 above)
Comment (ezrn) [reply champions/src/main/resources/configs/skills/skills.yml:593]: why

Comment (braulio-dev) [reply champions/src/main/resources/configs/skills/skills.yml:593]: consistency, every other skill has a default config. we also want to populate the entire skills.yml upon startup

Comment (ezrn) [reply champions/src/main/resources/configs/skills/skills.yml:593]: The in-class config IS the defaulty config, this config gets populated no matter what and just leads to unnecessary confusion because you have to separately change two configs

Comment (braulio-dev) [reply champions/src/main/resources/configs/skills/skills.yml:593]: kindos already told you to do it anyway just do it

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java (hunk 1 above)
Comment (ezrn) [reply champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java]: can get away with 25

File: champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java (hunk 4 above)
Comment (ezrn) [reply champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/data/BlockTossObject.java]: fixed anyways

**Comment by Mykindos** [comment]
@braulio-dev is this approved or what

**Comment by ezrn** [comment]
broken
